File: FoodDetectionApp/BarcodeScanner.swift
import Foundation
import Vision
import UIKit
import Combine

class BarcodeScanner: ObservableObject {
    @Published var lastBarcode: String?

    // MARK: - Barcode Detection

    func scanBarcode(from image: UIImage) {
        guard let cgImage = image.cgImage else {
            DispatchQueue.main.async { self.lastBarcode = nil }
            return
        }

        let request = VNDetectBarcodesRequest { [weak self] request, error in
            guard error == nil,
                  let results = request.results as? [VNBarcodeObservation] else {
                DispatchQueue.main.async { self?.lastBarcode = nil }
                return
            }

            // Look for EAN-13 or UPC-A barcodes
            let barcode = results.first(where: { (obs: VNBarcodeObservation) -> Bool in
                obs.symbology == VNBarcodeSymbology.ean13 || obs.symbology == VNBarcodeSymbology.ean8 || obs.symbology == VNBarcodeSymbology.upce
            })

            DispatchQueue.main.async {
                self?.lastBarcode = barcode?.payloadStringValue
            }
        }

        // Limit to the barcode symbologies we care about
        request.symbologies = [VNBarcodeSymbology.ean13, VNBarcodeSymbology.ean8, VNBarcodeSymbology.upce]

        let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
        DispatchQueue.global(qos: .userInitiated).async {
            try? handler.perform([request])
        }
    }

    // MARK: - Open Food Facts Lookup

    func lookupBarcode(_ code: String) async throws -> (name: String, info: NutritionInfo)? {
        guard let url = URL(string: "https://world.openfoodfacts.org/api/v0/product/\(code).json") else {
            return nil
        }

        let (data, _) = try await URLSession.shared.data(from: url)

        guard let root = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }

        // Check if the product was found
        guard let status = root["status"] as? Int, status == 1 else {
            return nil
        }

        guard let product = root["product"] as? [String: Any] else {
            return nil
        }

        let productName = product["product_name"] as? String ?? "Unknown Product"

        guard let nutriments = product["nutriments"] as? [String: Any] else {
            return nil
        }

        let calories = nutrimentValue(nutriments, key: "energy-kcal_100g")
        let carbs = nutrimentValue(nutriments, key: "carbohydrates_100g")
        let protein = nutrimentValue(nutriments, key: "proteins_100g")
        let fat = nutrimentValue(nutriments, key: "fat_100g")

        let info = NutritionInfo(
            calories: calories,
            recipe: "Scanned from barcode. Consider pairing with vegetables for a balanced meal.",
            carbs: carbs,
            protein: protein,
            fats: fat,
            source: "Open Food Facts",
            micros: nil
        )

        return (name: productName, info: info)
    }

    // MARK: - Helpers

    private func nutrimentValue(_ nutriments: [String: Any], key: String) -> String {
        if let val = nutriments[key] as? Double {
            return String(format: "%.1f", val)
        }
        if let val = nutriments[key] as? Int {
            return String(val)
        }
        if let val = nutriments[key] as? String {
            return val
        }
        return "0"
    }
}

File: .claude/worktrees/agent-ac129a56/android/app/src/main/java/com/foodsense/android/services/BarcodeScanner.kt
package com.foodsense.android.services

import android.graphics.Bitmap
import com.foodsense.android.data.NutritionInfo
import com.google.mlkit.vision.barcode.BarcodeScannerOptions
import com.google.mlkit.vision.barcode.BarcodeScanning
import com.google.mlkit.vision.barcode.common.Barcode
import com.google.mlkit.vision.common.InputImage
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.OkHttpClient
import okhttp3.Request
import java.util.concurrent.TimeUnit
import kotlin.coroutines.resume

class BarcodeScanner {

    private val options = BarcodeScannerOptions.Builder()
        .setBarcodeFormats(Barcode.FORMAT_EAN_13, Barcode.FORMAT_UPC_A)
        .build()

    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .build()

    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
    }

    suspend fun scanBarcode(bitmap: Bitmap): String? {
        val image = InputImage.fromBitmap(bitmap, 0)
        val scanner = BarcodeScanning.getClient(options)

        return suspendCancellableCoroutine { continuation ->
            scanner.process(image)
                .addOnSuccessListener { barcodes ->
                    val barcode = barcodes.firstOrNull {
                        it.format == Barcode.FORMAT_EAN_13 || it.format == Barcode.FORMAT_UPC_A
                    }
                    continuation.resume(barcode?.rawValue)
                }
                .addOnFailureListener {
                    continuation.resume(null)
                }
                .addOnCompleteListener {
                    scanner.close()
                }

            continuation.invokeOnCancellation {
                scanner.close()
            }
        }
    }

    suspend fun lookupBarcode(code: String): Pair<String, NutritionInfo>? = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("https://world.openfoodfacts.org/api/v0/product/$code.json")
            .get()
            .build()

        val response = runCatching {
            client.newCall(request).execute()
        }.getOrNull() ?: return@withContext null

        val body = response.body?.string() ?: return@withContext null
        if (!response.isSuccessful) return@withContext null

        val root = runCatching {
            json.parseToJsonElement(body).jsonObject
        }.getOrNull() ?: return@withContext null

        val status = root["status"]?.jsonPrimitive?.intOrNull
        if (status != 1) return@withContext null

        val product = root["product"]?.jsonObject ?: return@withContext null
        val productName = product["product_name"]?.jsonPrimitive?.contentOrNull ?: "Unknown Product"

        val nutriments = product["nutriments"]?.jsonObject ?: return@withContext null

        val calories = nutrimentValue(nutriments, "energy-kcal_100g")
        val carbs = nutrimentValue(nutriments, "carbohydrates_100g")
        val protein = nutrimentValue(nutriments, "proteins_100g")
        val fat = nutrimentValue(nutriments, "fat_100g")

        val info = NutritionInfo(
            calories = calories,
            recipe = "Scanned from barcode. Consider pairing with vegetables for a balanced meal.",
            carbs = carbs,
            protein = protein,
            fats = fat,
            source = "Open Food Facts",
            micros = null,
        )

        productName to info
    }

    private fun nutrimentValue(nutriments: JsonObject, key: String): String {
        val element = nutriments[key] ?: return "0"
        val primitive = element.jsonPrimitive
        primitive.doubleOrNull?.let { return String.format("%.1f", it) }
        primitive.intOrNull?.let { return it.toString() }
        primitive.contentOrNull?.let { return it }
        return "0"
    }
}
